Different direction

EECOLOR opened the pull request:
`routeMap` as the only option

When checking the code, you only need to look at the `examples`. And note that for the `advanced` version I did not get around to implementing `data` retrieval.

EECOLOR committed:
Refactor and add tests

Trying a different route ;-) where `routeMap` is the only option.

EECOLOR committed:
Move towards routeMap only routing

It seems better to only have 1 way of doing things. This simplifies code
and provides extra featues (fallback to not found, even after dynamic
routes).

It also provides some other benefits because we can have absolute paths
in a relative fashion.

EECOLOR committed:
Prefil available params in reverse routes

EECOLOR committed:
Internal refactoring

EECOLOR committed:
Fix advanced example

Incidentally added language support.
Some features still missing.
Needs a bit of polish, but ran out of time

EECOLOR commented:
When checking the code, you only need to look at the `examples`. And note that for the `advanced` version I did not get around to implementing `data` retrieval.

peeke commented on example/src/App.js:
```diff
@@ -1,78 +1,85 @@
+        ) || 'not matched'}
       </div>
-      {route('tab1', <div>Side bar for tab 1</div>)}
+      {route(tab1, <div>Side bar for tab 1</div>)}
```
Los van dit PR: volgensmij zou je hier altijd een functie willen. Als je hier een pagina hebt staan, wordt hier de virtual DOM al van opgebouwd, terwijl dat pas nodig is als de route matched.

EECOLOR replied:
Niet helemaal. Alleen datgene wat je hier ziet. Er wordt alleen `React.createElement` aangeroepen. Als dat een custom component is geef je alleen de functie mee, maar die wordt nog niet uitgevoerd.

Voor zover ik weet wordt de rest pas uitgevoerd bij een daadwerkelijke render.

EECOLOR replied:
In dit voorbeeld komt de `console.log` niet voorbij:
```js
function App() {
  const x = <Test />

  return 'test'
}

function Test() {
  console.log('render Test')
  return null
}
```

peeke replied:
Ah, right! Dan heb ik niks gezegd

peeke commented on example/src/advanced/App.js:
```diff
@@ -71,6 +56,10 @@ function Language({ children, basePath }) {
+
+  function setLanguage(language) {
+    navigate(currentRoute({ language }))
+  }
```
Wat hier in de praktijk zou kunnen gebeuren: pagina met slug `/nl/artikel/nederlandse-slug` wordt vertaald door `/en/artikel/engelse-slug`. Daardoor is het in sommige gevallen niet genoeg om slechts het language slug deel in de url te replacen. 

Heb nog niet echt een oplossing, maar misschien dat we het voorbeeld daar wel op aan zouden moeten passen. Je zou evt. kunnen zeggen dat die url redirect naar de url met de juiste slug, of die slug ergens ophaalt.

EECOLOR replied:
Ik zal het voorbeeld aanpassen zodra ik in de data verschillende slugs heb. Goed punt in ieder geval.

EECOLOR replied:
Ik pak dit later op

EECOLOR replied:
Hmm, ik heb nog even nagedacht. En ik denk dat het slimmer is om dit bij de page load te doen (op dezelfde plek als de afhandeling van permalinks). Het is namelijk superhandig als je via de adresbalk van taal kan wisselen. Moeten we nog maar een keer naar kijken of dat ook echt handig is.

EECOLOR replied:
Misschien sowieso slim om ook het alternatief (wel de goeie wissel naar juiste url) uit te voeren op basis van artikel in andere taal. Dan kunnen we meteen zien wat het verschil is in complexiteit en waar we die het liefst willen hebben.

peeke commented on example/src/advanced/App.js:
```diff
@@ -80,68 +69,70 @@ function useLanguage() {
-      <Link to={path.articles()[language]}>{path.articles.meta.title[language]}</Link>
-      <Link to={path.articles.article({ articleId: 'article1' })[language]}>Featured article</Link>
+      <Link to={home()}>Home</Link>
+      <Link to={articles()}>{articles[routeSymbol].data.title[language]}</Link>
```
Ergens zou ik het misschien handiger vinden om iets van  `getRoute(articles).data.title[langage]` te doen, dan hoef je buiten de library niet van het bestaan van het routeSymbol te weten (of te weten wat is een symbol is/waarom dat hier handig is). 

Of:
```const { title } = useRouteData(articles)```

EECOLOR replied:
Had hier nog niet echt aandacht aan besteed, focus was voornamelijk op de routes zelf. Als ik naar de data ga kijken zal ik checken hoe we dat het beste kunnen doen.

EECOLOR replied:
@peeke `data` is nu een 'non-enumerable' property van een `Route`

peeke commented on example/src/advanced/App.js:
```diff
@@ -80,68 +69,70 @@ function useLanguage() {
 
-  const knownPaths = [path.main, path.tab1, path.tab2]
-  const atValidTab = relativePick(...knownPaths.map(x => [x, true]))
+  const knownPaths = [main, tab1, tab2]
+  const atValidTab = pick(...knownPaths.map(x => [x, true]))
```
Dit zou wellicht ook nog kunnen:
```
const match = pick(main, tab1, tab2)
const atValidTab = Boolean(match)
...
<Tab selected={match === tab1} to={tab1()}>Tab1</Tab>
```

EECOLOR replied:
Ik zal even kijken of het handig is om daar de `pick` voor te gebruiken of dat het slimmer is om er een aparte functie voor te maken.

Zou ook kunnen dat een ander mechanisme juist weer handiger is, ik moet de code achter de schermen nog opruimen.

peeke commented on example/src/advanced/index.html.js:
```diff
@@ -4,20 +4,23 @@ import stylesheet from '@kaliber/build/lib/stylesheet'
-      [routeMap.notFound, { status: 404 }],
-      [routeMap.articles.article.notFound, { status: 404 }]
+  match(location, req) {
+    const language = req.acceptsLanguages('nl', 'en') || 'nl'
```
Cool, dit ken ik niet.

EECOLOR replied:
Ik ook niet, haha, maar nu kennen we het allebei :-D

peeke reviewed (commented):
Ik vind het een hele lekkere slag :+1: Ik heb wel comments, maar volgensmij gaan die eigenlijk geen van allen direct over de changes in dit PR ^^'

EECOLOR committed:
Correctly handle localization errors

Some refactoring as well

EECOLOR committed:
Working towards useful types

EECOLOR committed:
Refactor inner workings

- Group context by rate of change
- Only a single match for each location change
- More sensible hooks

EECOLOR committed:
Add support for route data

Also in this pull request:
- More advanced type definitions for routeMap
- Refactoring of route definition

EECOLOR committed:
Typescript related tweaks

EECOLOR committed:
Small simplification in example
